File: 2024/day12/part1.cpp
#include <iostream>
#include <map>
#include <numeric>
#include <vector>

using namespace std;

struct UnionFind {
    vector<int> p;

    UnionFind(int n) {
        p.resize(n);
        iota(begin(p), end(p), 0);
    }

    int find(int i) {
        if (p[i] == i) return i;
        return p[i] = find(p[i]);
    }

    bool join(int i, int j) {
        i = find(i);
        j = find(j);
        if (i == j) return false;
        p[i] = j;
        return true;
    }
};

int main() {
    string line;
    vector<string> grid;
    int n, m;
    while (getline(cin, line)) {
        grid.push_back(line);
    }
    n = grid.size();
    m = grid[0].size();

    UnionFind uf(n * m);

    for (int i = 0; i < n; ++i) for (int j = 0; j < m; ++j) {
        if (i > 0 && grid[i-1][j] == grid[i][j]) {
            uf.join(i * m + j, (i - 1) * m + j);
        }

        if (j > 0 && grid[i][j - 1] == grid[i][j]) {
            uf.join(i * m + j, i * m + j - 1);
        }
    }

    map<int, vector<int>> groups;

    for (int i = 0; i < n * m; ++i) {
        groups[uf.find(i)].push_back(i);
    }

    int ans = 0;
    for (auto& [k, v] : groups) {
        int area = v.size();
        int per = 0;
        for (int idx : v) {
            int i = idx / m;
            int j = idx % m;

            for (auto [ni, nj] : vector<pair<int,int>>{{i-1,j},{i+1,j},{i,j-1},{i,j+1}}) {
                if (ni < 0 || ni >= n || nj < 0 || nj >= m) {
                    ++per;
                    continue;
                }

                per += (grid[ni][nj] != grid[i][j]);
            }
        }
        ans += per * area;
    }
    cout << ans << endl;
}

File: 2024/day14/part1.lua

function Main()
    local n = 103
    local m = 101
    local mt = {}
    for i=1,2 do
        mt[i] = {}
        for j=1,2 do
            mt[i][j] = 0
        end
    end
    for line in io.lines() do
        local _, _, px, py, vx, vy = string.find(line, "p=(-?%d+),(-?%d+)%sv=(-?%d+),(-?%d+)")
        px = tonumber(px)
        py = tonumber(py)
        vx = tonumber(vx)
        vy = tonumber(vy)

        px = px + 100 * vx
        py = py + 100 * vy

        px = px % m
        py = py % n

        if px == (m >> 1) then
            goto continue
        end
        if py == (n >> 1) then
            goto continue
        end

        local i = 1
        local j = 1

        if px > (m >> 1) then
            j = j + 1
        end

        if py > (n >> 1) then 
            i = i + 1
        end

        mt[i][j] = mt[i][j] + 1

        ::continue::
    end

    local ans = mt[1][1] * mt[1][2] * mt[2][1] * mt[2][2]

    print(ans)

end

Main()
